[mlir][OpenMP] inscan reduction modifier and scan op mlir support

anchuraj opened the pull request:
Scan directive allows to specify scan reductions within an worksharing loop, worksharing loop simd or simd directive which should have an `InScan` modifier associated with it. This change adds the mlir support for the same.

Related PR: [Parsing and Semantic Support for scan](https://github.com/llvm/llvm-project/pull/102792)

anchuraj committed:
Changes for inscan reduction and scan op

anchuraj committed:
R2: Addressing a few review comments

anchuraj commented on mlir/include/mlir/Dialect/OpenMP/OpenMPEnums.td:
```diff
@@ -178,6 +178,27 @@ def OrderModifier
+def ReductionModifierInScan : I32EnumAttrCase<"InScan", 0>;
+def ReductionModifierTask : I32EnumAttrCase<"Task", 1>;
+def ReductionModifierDefault : I32EnumAttrCase<"Default", 2>;
+
```
Thank you! I have updated

anchuraj commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -1958,7 +2001,7 @@ void WsloopOp::build(OpBuilder &builder, OperationState &state,
         /*nowait=*/false, /*order=*/nullptr, /*order_mod=*/nullptr,
         /*ordered=*/nullptr, /*private_vars=*/{}, /*private_syms=*/nullptr,
-        /*reduction_vars=*/ValueRange(), /*reduction_byref=*/nullptr,
+        nullptr, /*reduction_vars=*/ValueRange(), /*reduction_byref=*/nullptr,
```
Thank you for the suggestion. Updated.

anchuraj commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -2840,6 +2886,54 @@ void MaskedOp::build(OpBuilder &builder, OperationState &state,
+    }
+  }
+  return emitError("Scan Operation should be enclosed within a parent "
+                   "WORSKSHARING LOOP or SIMD with INSCAN reduction modifier");
```
Updated.

mjklemm reviewed (approved):
LGTM

skatrak reviewed (commented):
Thank you Anchu for working on this, I have some small implementation comments, but I also have some concerns about your proposed representation for `omp.scan`.

Other than that, I'd also suggest adding a not-yet-implemented error in MLIR to LLVM IR translation for the new reduction modifier attributes, and update mlir/test/Target/LLVMIR/openmp-todo.mlir to check for those and for the error that should already be triggered for `omp.scan`.

anchuraj commented on mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:
```diff
@@ -455,6 +523,7 @@ class OpenMP_InReductionClauseSkip<
   ];
 
   let arguments = (ins
+    OptionalAttr<ReductionModifierAttr>:$in_reduction_mod,
```
Hi Sergio, I followed this initially. However, checking the spec, I see modifier listed for all reduction though their behavior may not be fully specified. I decided to stick on to the spec.

skatrak replied:
I may be misunderstanding your concern, so let me know if I am, but are you referring to the _reduction-identifier_ in the _Modifiers_ table? That would be e.g. the '+' in `reduction(+: acc)`, which is indirectly represented by the `*_syms` operands.

The _reduction-modifier_ that you're introducing a new operand for is actually only present in the _Modifiers_ table for `reduction`, as far as I can tell.

anchuraj commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -517,14 +523,25 @@ static ParseResult parseClauseWithRegionArgs(
     return failure();
 
+  StringRef enumStr;
+  if (succeeded(parser.parseOptionalKeyword("Id"))) {
+    if (parser.parseColon() || parser.parseKeyword(&enumStr) ||
+        parser.parseComma())
+      return failure();
+    std::optional<ReductionModifier> enumValue =
+        symbolizeReductionModifier(enumStr);
+    *reductionMod = ReductionModifierAttr::get(parser.getContext(), *enumValue);
+  }
```
Sounds good. The name of the modifier is `reduction identifier` in spec. That is the reason why I chose `id`

anchuraj commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:
```diff
@@ -1560,6 +1560,26 @@ def CancellationPointOp : OpenMP_Op<"cancellation_point", clauses = [
   let hasVerifier = 1;
 }
 
+def ScanOp : OpenMP_Op<"scan", [
```
Great question. There are a few facts to be considered here. For Simplicity, i will consider body of for-loop as 
```
{
  <pre scan block> b1;
  scan
  <post scan block> b2;
} 
```
- As per the definition of scan,  `b1` or `b2` can be input or scan phase based on `exclusive` or `inclusive` clause used with the directive
- My first attempt was to define this structure in parsing, however I failed  `<structured sequence> scan <structured sequence>` was making the grammar ambiguous. I digged into clang and saw that the blocks are not associated with the directive (scan is a stand-alone directive). Its the llvm lowering that separated the two blocks. When lowered, the body of the for loop before scan is emitted as first basic block and after scan is emitted as another basic block. On encountering scan, these blocks are treated as input or scan phase based on clauses that scan has. Since there are only two blocks and as scan directive always appear in loops with `inscan` reduction modifier, dealing it while lowering would be simple.
- When parsing does not separate it easily, representing it as an MLIR op will be very complicated to do in frontend and would result in code that is not much reusable if we go with that representation. 

These are why the current representation is chosen.

skatrak replied:
My understanding is that clang would be able to base the implementation of `scan` on the position of the directive within the loop body (just like how it's defined at the source level by the spec) because the AST is not modified by moving statements around possibly crossing the `scan` split point. Which would also be the reason why you are able to parse it that way in Flang.

However, MLIR transformation passes are able to move operations within a region in a way that would potentially break `scan`. We must make sure the MLIR representation we choose prevents this from happening, and at the moment the best I can think of would be the 2-region alternative I illustrated above.

But I see the problem you point to about the naming of these regions. However, if instead of calling them `input` and `scan`, we call them `pre` and `post` then we should be able to lower them to LLVM IR properly based on whether it's an `inclusive` or `exclusive` scan. That would mean something like this:
```mlir
omp.wsloop reduction(inscan @reduction %0 -> %red0 : !llvm.ptr) {
  omp.loop_nest ... {
    omp.scan exclusive(%red0 : !llvm.ptr)
    pre {
      // Everything before '!$omp scan ...'
      ...
      omp.terminator
    } post {
      // Everything after '!$omp scan ...'
      ...
      omp.terminator
    }
    omp.yield
  }
}
```
The questions that I guess would remain for this approach would be whether we can let common sub-expressions, constants etc. be hoisted out of the `pre` and `post` regions, whether we should allow other non-terminator operations after `omp.scan` inside of the loop and whether we need to make values defined in `pre` accessible to `post` by having an `omp.yield` terminator for the former and corresponding entry block arguments for the latter.

CC: @kiranchandramohan, @tblah.

skatrak replied:
Thank you @tblah and @kiranchandramohan for sharing your thoughts. @anchuraj and I had a call about this the other day, and at the time the 2-region approach made the most sense. But if we can somehow ensure that no MLIR transformations are able to move operations across the `scan` split point in a way that breaks semantics, it is my understanding that the translation to LLVM IR wouldn't be much more difficult than having explicit regions, and it would potentially make the representation and Flang lowering simpler in general.

With regards to both approaches introducing regions, they have the disadvantage that the current to the PFT to MLIR lowering system doesn't seem well-suited to lower only parts of a block of code, so we thought that, if we followed that approach, then we'd probably want to first produce an `omp.scan` with both regions empty at first, while lowering the rest of the loop normally. After that, we'd introduce another Flang-specific MLIR pass (similar to function filtering, map info finalization, etc.) that would sink operations before the `omp.scan` into its `pre` region and operations after it into its `post` region. If there were values defined in `pre` used in `post`, that pass would be able to detect them and produce the corresponding `omp.yield` and `post` entry block arguments.

If a common subexpression doesn't depend on the loop index, I agree that this would result in no problems for any of the approaches, since it should be hoisted out of the loop body. Same thing with constants, for example. However, one case I had in mind, which I think could potentially result in MLIR optimizations moving operations across the split point would be the `i + 1` expression below:
```f90
subroutine foo(v1, v2, size)
  implicit none
  integer, intent(in) :: size, v1(size)
  integer, intent(out) :: v2(size)
  integer :: sum, i
  sum = 0
  !$omp parallel do reduction(+:sum)
  do i = 0, size - 1
    sum = sum + v1(i + 1)
    !$omp scan inclusive(sum)
    v2(i + 1) = sum
  end do
  !$omp end parallel do
end subroutine foo
```

At the moment that results in separate `fir.load`s for `i` and separate `arith.constant`s for `1`, so the only values that would be used both inside of `pre` and `post` would be the `hlfir.declare` for the `sum` reduction variable (which we may want to leave out of both regions or put it in `pre`). But I can't see any reason why, if we don't split the loop body, MLIR passes wouldn't be able to realize there are 2 loads to a same variable that is not modified and calculate `i + 1` only once at the beginning. That kind of transformation doesn't seem like it would break the region-less approach, but it would be one case introducing the need for passing values from `pre` to `post`.

So I'm thinking that perhaps it would indeed be enough to mark `omp.scan` as reading from and writing to its `inclusive / exclusive` arguments. That should ensure no ops related to these values will be reordered across the split point, and that `omp.scan` won't be itself moved in a way that results in such a reordering. Maybe that's enough to ensure valid passes won't break this operation.

anchuraj replied:
Thank you for your suggestions @tblah, @skatrak and @kiranchandramohan. 
Looking at the standard, following is a restriction on scan directive. 
```
Intra-iteration dependences from a statement in the structured block sequence that precede a scan directive to a statement in the structured block sequence that follows a scan directive must not exist, except for dependences for the list items specified in an inclusive or exclusive clause.
```
The above restriction makes the examples like the one using `block` construct mentioned by @kiranchandramohan  invalid.
From  the example @skatrak used, it transforms program to a way where yielding would be required. It transforms program to a structure which is not compliant with the above OpenMP restriction ( which need not be hold at the MLIR transformation stage) 

Circling back on @kiranchandramohan's question, is there a way to prevent such optimizations inside `WsloopOp` with `inscan` modifier, considering `scan` as side-effecting?  Please  let me know @skatrak  and @tblah your thoughts.

tblah replied:
Yeah I guess giving the scan memory effects could work. I can't think of a reason off the top of my head why @skatrak's suggestion to use the memory effects only on the exclusive/inclusive arguments wouldn't work.

But just to be safe I would probably do the initial implementation saying it has some broadly specified memory effects. If we follow the recommendations for modelling syscalls given [here](https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/), I think that should ensure that any observable effects before the scan must stay before (and vice versa). 

So it would be modeled a bit like

```
input phase code;
write(1, "I am a scan statement.\n", 23);
output phase code;
```

anchuraj replied:
Thank you. I discussed this offline with @skatrak. Based on the discussions and inputs, I am proceeding with implementing no region approach with `ScanOp` having memory side effects . Please let me know if there are any further concerns on this.

anchuraj replied:
Updated the operation with `MemWrite` effects

anchuraj commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:
```diff
@@ -1560,6 +1560,26 @@ def CancellationPointOp : OpenMP_Op<"cancellation_point", clauses = [
 }
 
+def ScanOp : OpenMP_Op<"scan", [
+    AttrSizedOperandSegments, RecipeInterface, IsolatedFromAbove
```
I will remove it. Those were accidental.

anchuraj committed:
qR3: Addressing a few review comments

anchuraj committed:
R4: Adding memory write side effects to ScanOp and addressing other review comments

anchuraj commented on mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:
```diff
@@ -1008,6 +1082,7 @@ class OpenMP_ReductionClauseSkip<
   ];
 
   let arguments = (ins
+    OptionalAttr<ReductionModifierAttr>:$reduction_mod,
```
Updated

anchuraj commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -2840,6 +2887,55 @@ void MaskedOp::build(OpBuilder &builder, OperationState &state,
+    if (parentOp.getReductionModAttr() &&
+        parentOp.getReductionModAttr().getValue() ==
+            mlir::omp::ReductionModifier::InScan) {
+      if (!verifyScanVarsInReduction(parentOp.getReductionVars())) {
```
Thank you for the correction. Updated.

anchuraj committed:
R5: Removing the check for matching scan variables and reduction variables

anchuraj commented:
I have removed the check to compare `scanVars` and `reductionBlockArgs` as mlir phase inserts a new copy of private argument inside the loop body and that is the value used in `scanVars` (as shown in the below code snippet). This check is performed in the semantic phase in `flang` ensuring that there will not be any errors from the input code.

```
     omp.wsloop reduction(mod: inscan, @add_reduction_i32 %11 -> %arg0 : !fir.ref<i32>) {
        omp.loop_nest (%arg1) : i32 = (%c1_i32) to (%c100_i32) inclusive step (%c1_i32) {
          %45 = fir.declare %arg0 {uniq_name = "_QFEx"} : (!fir.ref<i32>) -> !fir.ref<i32>
          ...
          omp.scan inclusive(%45 : !fir.ref<i32>)
         ...
```

skatrak commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -3125,6 +3153,41 @@ void MaskedOp::build(OpBuilder &builder, OperationState &state,
+    return emitError(
+        "Exactly one of EXCLUSIVE or INCLUSIVE clause is expected");
+  }
+  if (mlir::omp::WsloopOp parentWsLoopOp =
```
Nit: We're already in the `mlir::omp` namespace (and there's a `using mlir::omp` at the top of the file). There are a few more instances in this function.
```suggestion
  if (WsloopOp parentWsLoopOp =
```

skatrak commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -3125,6 +3153,41 @@ void MaskedOp::build(OpBuilder &builder, OperationState &state,
+}
+
+LogicalResult ScanOp::verify() {
+  if (hasExclusiveVars() && hasInclusiveVars()) {
```
Nit: Braces.

skatrak commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -3125,6 +3153,41 @@ void MaskedOp::build(OpBuilder &builder, OperationState &state,
+          (*this)->getParentOfType<mlir::omp::WsloopOp>()) {
+    if (parentWsLoopOp.getReductionModAttr() &&
+        parentWsLoopOp.getReductionModAttr().getValue() ==
+            mlir::omp::ReductionModifier::inscan) {
```
Nit: Braces.

skatrak commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -3125,6 +3153,41 @@ void MaskedOp::build(OpBuilder &builder, OperationState &state,
+          (*this)->getParentOfType<mlir::omp::SimdOp>()) {
+    if (parentSimdOp.getReductionModAttr() &&
+        parentSimdOp.getReductionModAttr().getValue() ==
+            mlir::omp::ReductionModifier::inscan) {
```
Nit: Braces.

skatrak reviewed (commented):
Thank you Anchu for this work. I think this representation should work, my remaining comments are mostly just about properly communicating currently unsupported cases to users and some minor nits.

anchuraj committed:
R6: Addressing review comments

anchuraj commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -4591,6 +4591,10 @@ convertHostOrTargetOperation(Operation *op, llvm::IRBuilderBase &builder,
+      .Case([&](omp::ScanOp) {
+        return op->emitError()
+               << "not yet implemented: " << op->getName() << " operation";
+      })
```
Thank you for the suggestion. Updated.

skatrak commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -615,15 +632,14 @@ static ParseResult parseBlockArgClause(
   }
   return success();
 }
-
```
Remember to add back this line before merging.

skatrak commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -3125,6 +3153,34 @@ void MaskedOp::build(OpBuilder &builder, OperationState &state,
+  if (hasExclusiveVars() == hasInclusiveVars())
+    return emitError(
+        "Exactly one of EXCLUSIVE or INCLUSIVE clause is expected");
+  if (WsloopOp parentWsLoopOp = (*this)->getParentOfType<WsloopOp>())
+    if (parentWsLoopOp.getReductionModAttr() &&
+        parentWsLoopOp.getReductionModAttr().getValue() ==
+            mlir::omp::ReductionModifier::inscan)
+      return success();
+  if (SimdOp parentSimdOp = (*this)->getParentOfType<SimdOp>())
+    if (parentSimdOp.getReductionModAttr() &&
+        parentSimdOp.getReductionModAttr().getValue() ==
+            mlir::omp::ReductionModifier::inscan)
+      return success();
```
Nit: Some more `mlir::omp` left behind, and braces are preferred for the outer `if` statements to avoid future "dangling else" issues while removing them for the inner statements ([link](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)). 
```suggestion
  if (WsloopOp parentWsLoopOp = (*this)->getParentOfType<WsloopOp>()) {
    if (parentWsLoopOp.getReductionModAttr() &&
        parentWsLoopOp.getReductionModAttr().getValue() ==
            ReductionModifier::inscan)
      return success();
  }
  if (SimdOp parentSimdOp = (*this)->getParentOfType<SimdOp>()) {
    if (parentSimdOp.getReductionModAttr() &&
        parentSimdOp.getReductionModAttr().getValue() ==
            ReductionModifier::inscan)
      return success();
  }
```

skatrak commented on mlir/test/Dialect/OpenMP/invalid.mlir:
```diff
@@ -1825,6 +1825,84 @@ func.func @omp_cancellationpoint2() {
+  %test1f32 = "test.f32"() : () -> (!llvm.ptr)
+  omp.taskloop reduction(mod:inscan, @add_f32 %test1f32 -> %arg1 : !llvm.ptr) {
+    omp.loop_nest (%i, %j) : i32 = (%lb, %ub) to (%ub, %lb) step (%step, %step) {
+  // expected-error @below {{SCAN directive needs to be enclosed within a parent worksharing loop construct or SIMD construct with INSCAN reduction modifier}}
```
Nit: Add a single test to check this error for the case where there's a parent `omp.wsloop` without `mod: inscan`.

skatrak reviewed (approved):
Thank you, this LGTM. Just minor comments, but no need for a second review by me before merging.

anchuraj committed:
R7: Addressing review comments

anchuraj commented:
> Thank you, this LGTM. Just minor comments, but no need for a second review by me before merging.

Thank you @skatrak for reviewing my PR! I have addressed the review comments
